snapshots/devmapper/pool_device.go

```diff
@@ -344,7 +344,7 @@ func (p *PoolDevice) CreateSnapshotDevice(ctx context.Context, deviceName string
 		log.G(ctx).Debugf("suspending %q before taking its snapshot", deviceName)
 		suspendErr := p.SuspendDevice(ctx, deviceName)
 		if suspendErr != nil {
-			return suspendErr
+			return multierror.Append(suspendErr, p.metadata.MarkFaulty(ctx, snapInfo.Name))
```

Comment:
I don't think there is a relation between device ID we allocated (but not yet used) and suspending the device, since these are completely independent (up until we take the actual snapshot). I feel like suspend is something that worth retrying in subsequent calls.

I'm also not quite sure in which state snapshot device should end up when we failed to suspend parent (mark faulty or rollback everything).

So what I'd like to propose is how do you feel if we try to suspend parent first, and then allocate device ID for child snap? (like move suspend up, to right after `baseInfo, err := p.metadata.GetDevice(ctx, deviceName)`). So we suspend, allocate device ID, take snapshot, etc.

Reply:
That makes perfect sense to me. I wasn't really sure what the right approach would be since I am not super familiar with the code - I just experienced the issue and figured that whatever is happening, it ends with the ID getting re-used and subsequent calls failing.  So I started with the thing that seemed like the least change.

I'll update my commit to move the suspend logic above the allocation.